os/drivers/pm/pm.c

```diff
@@ -0,0 +1,204 @@
+ *   the command. 
+ *   for LOCK, WAKEUP_TIMER, SET_WIFI_WAKEUP, arg should be a long unsigned type.
+ *   for TIMETICK , arg should be a pointer to a size_t type.
+ *   for TUNEFREQ, arg should be an int type.
+ *
+ * Description:
+ *   This api can be used to perform PM operation.
+ *
+ ************************************************************************************/
+static int pm_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	int ret = -ENOSYS;
+    
+	/* Handle built-in ioctl commands */
+	switch (cmd) {
+    
+     	/****************************************************************************
+     	* Description: Lock PM in PM_NORMAL state by invoking pm_stay().
+     	****************************************************************************/
+    	case PMIOC_LOCK:
+        	/* Always lock PM_NORMAL state, check again whether this implementation
+        	is enough to handle, if not, consider below methods...
+        	1a. PM_ACTIVITY is invoked in some of the driver, if application utlized it, it should draw state back to PM_NORMAL
+        	1b. If 1a cannot cover all condition, add one more syscall api, as a bridge for application to call pm_activity
+        	2a. Else, we can consider to lock/unlock for all state other than PM_SLEEP, ie. fetch state: pm_querystate(PM_IDLE_DOMAIN)
+        	2b. Need to verify if the transition is going to happen, but lock state appears suddenly, what will happen?
+        	*/
+        	pm_stay(PM_IDLE_DOMAIN, PM_NORMAL);
+        	fvdbg("State locked!\n");
+        	if (arg > 0) {
+            		pm_set_timer(PM_LOCK_TIMER, arg);
+        	} else {
+            		g_pm_timer.timer_type = PM_NO_TIMER;
+        	}
+        	ret = OK;
+        	break;
+        /****************************************************************************
+         * Description:
+         *  Unlock PM from PM_NORMAL state by invoking pm_relax(). 
+         ****************************************************************************/
+        case PMIOC_UNLOCK:
+            	pm_relax(PM_IDLE_DOMAIN, PM_NORMAL);
+            	fvdbg("State unlocked!\n");
+            	g_pm_timer.timer_type = PM_NO_TIMER;
+            	ret = OK;
+            	break;
+        /****************************************************************************
+         * Description:
+         *  Make the board wake up after given time (arg) when board sleeps.
+         ****************************************************************************/
+        case PMIOC_WAKEUP_TIMER:
+        	if (g_pm_timer.timer_type == PM_NO_TIMER && arg > 0) {
+        		pm_set_timer(PM_WAKEUP_TIMER, arg);
+        	}	 
+        	ret = OK;
+        	break;
+        /****************************************************************************
+         * Description:
+         *  Sets timer interval for next sleep cycle after sending of wifi keep 
+         *  alive signal from UART. It also stores the current time tick 
```

Comment:
"after sending of wifi keep alive signal from UART" can be an example of this command.
But it should not be a main content.